tools/convert_hf_to_sequential.py

```diff
@@ -0,0 +1,490 @@
+CUDA_VISIBLE_DEVICES=0,1,2,3 python ./deepy.py tools/convert_hf_to_sequential.py \
+    -d configs pythia/70M.yml local_setup.yml
+"""
+
+MULTI_GPU_ARGS = " ".join(
+        [
+            "--size 70m",
+            "--revision 143000",
+            "--output-dir checkpoints/neox_converted/pythia/70m",
+            "--cache-dir checkpoints/HF",
+            "--config configs/pythia/70M.yml configs/local_setup.yml",
+            "--test",
+            "--deduped",
+        ]
+    )
+
+
+def convert_hf_to_sequential(hf_model,seq_state_dict):
+    """Converts the weights of a HuggingFace model to neox 2.0 format.
+    
+    :param hf_model: the huggingface model
+    :param seq_state_dict: the state dict of the equivalent neox model
+    
+    returns the updated sequential state dict
+    """
+    num_layers = hf_model.config.num_hidden_layers
+    # Embedding is layer idx 0
+    seq_state_dict['sequential.0.word_embeddings.weight'] =\
+        hf_model.gpt_neox.embed_in.state_dict()['weight']
+    
+    for layer_hf in range(num_layers):
+        # offset by 2
+        layer_seq = layer_hf + 2
+
+        # get layer from hf model
+        hf_layer = hf_model.gpt_neox.layers[layer_hf]
+        hf_layer_sd = hf_layer.state_dict()
+        
+        
+        for key in hf_model.gpt_neox.layers[0].state_dict().keys():
+            
+            if key in ["attention.bias","attention.masked_bias"]:
+                continue
+            seq_state_dict[f"sequential.{layer_seq}.{key}"] = hf_layer_sd[key]
+            
+    # Load final layer norm
+    layer_seq = num_layers + 3
+    seq_state_dict[f"sequential.{layer_seq}.norm.weight"] = \
+        hf_model.gpt_neox.final_layer_norm.state_dict()['weight']
+    seq_state_dict[f"sequential.{layer_seq}.norm.bias"] = \
+        hf_model.gpt_neox.final_layer_norm.state_dict()['bias']
+    
+    # output embedding / LM head
+    layer_seq += 1
+    seq_state_dict[f"sequential.{layer_seq}.final_linear.weight"] = \
+        hf_model.embed_out.state_dict()['weight']
+
+
+
+def shard_sequential_mp(num_mp_ranks,sequential):
+    """Shards the sequential model into model parallel ranks.
+
+    :param num_mp_ranks: the number of model parallel ranks
+    :param sequential: the state dict of the sequential model at mp=1
+
+    returns a dict of state dicts for each mp rank
+    """
+    ranks = {x:dict() for x in range(num_mp_ranks)}
+    for k,v in sequential.items():
+        if reduce(np.logical_or,[x in k for x in ['layernorm', 
+                                                  'rotary_emb',
+                                                  'dense_4h_to_h.bias',
+                                                  'norm.weight',
+                                                  'norm.bias',
+                                                  'attention.dense.bias']]):
+            # no splitting
+            for x in range(num_mp_ranks):
+                ranks[x][k] = v
+        else:
```

Comment:
I think here ColumnParallel and RowParallel layers would have to be handled separately:

For example linear layer for query_key_value is difference from the dense in attention
https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/megatron/model/transformer.py#L221

https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/megatron/model/transformer.py#L305

Same goes for the two linear layers in MLP in each transformer layer.
https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/megatron/model/transformer.py#L96
https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/megatron/model/transformer.py#L106


As reference: The conversion scripts from 1.0 and current version to HF are handling it separately:

RowParallel Linear Layers:
https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/tools/convert_sequential_to_hf.py#L188-L194

ColumnParallel Linear Layers:
https://github.com/EleutherAI/gpt-neox/blob/586f5149ea948cc46b270ddb9e8076628868c34a/tools/convert_sequential_to_hf.py#L208-L215

Reply:
So currently my script handles column and row parallel by assuming that the larger dimension is the one to be shared. Since the non-sharded dimension is typically the hidden state, this could cause problems for larger models. Let me fix this